70_FLIPFluids/gpuRadixSort.h

```diff
@@ -0,0 +1,470 @@
+				shaderSrc->setShaderStage(IShader::E_SHADER_STAGE::ESS_COMPUTE);
+				if (!shaderSrc)
+					return nullptr;
+
+				return m_device->createShader(shaderSrc.get());
+			};
+
+		{
+            smart_refctd_ptr<video::IGPUShader> shader = compileShader("app_resources/compute/radix_sort/buildHistogram.comp.hlsl");
+
+            IGPUComputePipeline::SCreationParams params = {};
+			params.layout = histogramPplnLayout.get();
+            params.shader.shader = shader.get();
+
+            m_device->createComputePipelines(nullptr, {&params, 1}, &m_buildHistogramPipeline);
+        }
+		{
+            smart_refctd_ptr<video::IGPUShader> shader = compileShader("app_resources/compute/radix_sort/scanSpine.comp.hlsl");
+
+            IGPUComputePipeline::SCreationParams params = {};
+			params.layout = spinePplnLayout.get();
+            params.shader.shader = shader.get();
+
+            m_device->createComputePipelines(nullptr, {&params, 1}, &m_scanSpinePipeline);
+        }
+        {
+            smart_refctd_ptr<video::IGPUShader> shader = compileShader("app_resources/compute/radix_sort/radixSort.comp.hlsl");
+
+            IGPUComputePipeline::SCreationParams params = {};
+			params.layout = sortPplnLayout.get();
+            params.shader.shader = shader.get();
+
+            m_device->createComputePipelines(nullptr, {&params, 1}, &m_radixSortPipeline);
+        }
+
+        std::array<IGPUDescriptorSetLayout*, 2> dscLayoutPtrs = {
+				nullptr,
+				histDsLayout.get()
+			};
+		const uint32_t setCounts[2u] = { 0u, 2u };
+        histogramDsPool = m_device->createDescriptorPoolForDSLayouts(IDescriptorPool::ECF_UPDATE_AFTER_BIND_BIT, std::span(dscLayoutPtrs.begin(), dscLayoutPtrs.end()), setCounts);
+		m_histogramDs[0] = histogramDsPool->createDescriptorSet(histDsLayout);
+		m_histogramDs[1] = histogramDsPool->createDescriptorSet(histDsLayout);
+
+		dscLayoutPtrs = {
+				nullptr,
+				sortDsLayout.get()
+			};
+		sortDsPool = m_device->createDescriptorPoolForDSLayouts(IDescriptorPool::ECF_UPDATE_AFTER_BIND_BIT, std::span(dscLayoutPtrs.begin(), dscLayoutPtrs.end()), setCounts);
+		m_radixSortDs[0] = sortDsPool->createDescriptorSet(sortDsLayout);
+		m_radixSortDs[1] = sortDsPool->createDescriptorSet(sortDsLayout);
+
+		dscLayoutPtrs = {
+				nullptr,
+				spineDsLayout.get()
+			};
+		m_scanSpineDsPool = m_device->createDescriptorPoolForDSLayouts(IDescriptorPool::ECF_UPDATE_AFTER_BIND_BIT, std::span(dscLayoutPtrs.begin(), dscLayoutPtrs.end()));
+		m_scanSpineDs = m_scanSpineDsPool->createDescriptorSet(spineDsLayout);
+    }
+
+    void sort(IGPUCommandBuffer* cmdbuf, smart_refctd_ptr<IGPUBuffer> dataBuffer, uint32_t numElements, bool shouldWriteDs = false)
+    {
+		uint32_t partitionSize = NUM_PARTITIONS * SORT_WORKGROUP_SIZE;
+		uint32_t numWorkgroups = (numElements + partitionSize - 1) / partitionSize;
+
+        shouldWriteDs = shouldWriteDs || updateBuffers(numElements, numWorkgroups, dataBuffer->getSize() / numElements);
```

Comment:
one would usually take a `SBufferRange` instead of a buffer for a generic utility (people like to keep multiple things in their buffers) so you can sort a subrange